### PR TITLE
Allow volume attach limit overwrite via command line parameter

### PR DESCRIPTION
**Pull request by rfranzke**
**Is this a bug fix or adding new feature?**
/kind feature

**What is this PR about? / Why do we need it?**
This PR is a part of https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/347 and proceeds with the way suggested in https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/347#issuecomment-638177835: In order to allow existing clusters that are leveraging/relying on the `KUBE_MAX_PD_VOLS` feature for in-tree provisioners in Kubernetes (https://kubernetes.io/docs/concepts/storage/storage-limits/#custom-limits) to migrate to CSI, the EBS CSI driver is supporting the `--volume-attach-limit-overwrite` flag now until a more sophisticated solution can be implemented.

PS: I was adding a few missing doc strings along the way and reduced package/function name stuttering according to the [official Go suggestions](https://blog.golang.org/package-names). Hope that's fine.

### CODE REVIEW

**Comment by rfranzke** [comment]
cc @gnufied

**Comment by hardikdr** [comment]
@leakingtapan @wongma7 It would be helpful if you could also take a quick look.

It seems like a reasonable change, to not break the user experience with CSI migration on AWS/EKS. Basically the `KUBE_MAX_PD_VOLS` on kube-scheduler  is ineffective with CSI-drivers and leaves no way for end-user to overwrite  max-attachable-volumes. 
- This PR at least helps in keeping existing experience intact.

**Comment by wongma7** [comment]
Giving @leakingtapan a chance to chime in as he had some objections to this but I agree it's necessary to make the limit configurable somehow and a flag seems the best way.

**Comment by rfranzke** [comment]
@leakingtapan Do you have some time to check the PR? Any thoughts on it?

**Comment by rfranzke** [comment]
Hm, what's the general review process now? As @leakingtapan seems to be busy/inactive, can another owner/approver help to progress with this PR? @bertinatto @jsafrane @gnufied

File: cmd/options/node_options.go [hunk 1]
```diff
@@ -21,6 +21,15 @@ import (
...
+	// in CSINode objects. It is similar to https://kubernetes.io/docs/concepts/storage/storage-limits/#custom-limits
+	// which allowed administrators to specify custom volume limits. Note that this solution is just a "stop-gap"
+	// solution until a more sophisticated solution can be implemented (dynamically discovering the maximum number of
+	// attachable volume per EC2 machine type, see also https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/347).
+	VolumeAttachLimitOverwrite int64
+}
 
-func (s *NodeOptions) AddFlags(fs *flag.FlagSet) {}
+func (o *NodeOptions) AddFlags(fs *flag.FlagSet) {
+	fs.Int64Var(&o.VolumeAttachLimitOverwrite, "volume-attach-limit-overwrite", -1, "Overwrite value for the maximum number of attachable volumes for all nodes")
```
Comment (gnufied) [inline cmd/options/node_options.go:34]: I would prefer this is simply called `volume-attach-limit` integer, so as if not specified the documentation will say - value is derived from instance type.

File: cmd/options/server_options.go [hunk 2]
```diff
@@ -28,6 +28,6 @@ type ServerOptions struct {
 	Endpoint string
 }
 
-func (s *ServerOptions) AddFlags(fs *flag.FlagSet) {
-	fs.StringVar(&s.Endpoint, "endpoint", driver.DefaultCSIEndpoint, "Endpoint for the CSI driver server")
+func (o *ServerOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.Endpoint, "endpoint", driver.DefaultCSIEndpoint, "Endpoint for the CSI driver server")
```
Comment (gnufied) [inline cmd/options/server_options.go:32]: same as above. I am not sure why this change was done.

File: pkg/driver/driver.go [hunk 3]
```diff
@@ -40,37 +40,43 @@ const (
...
-	options *DriverOptions
+	options *Options
 }
 
-type DriverOptions struct {
-	endpoint        string
-	extraVolumeTags map[string]string
-	mode            Mode
+// Options contains configuration options for the driver.
+type Options struct {
```
Comment (gnufied) [inline pkg/driver/driver.go:59]: Just checking was `DriverOptions` to `Options` change made to make `govet` etc happy?  having this change included in the same commit makes it harder to review the actual code that this PR proposes. If you really want to make this change then please do that in a different commit.

**Comment by rfranzke** [comment]
@gnufied I've dropped all unrelated changes.

File: cmd/options/node_options.go [hunk 4]
```diff
@@ -21,6 +21,15 @@ import (
 )
 
 // NodeOptions contains options and configuration settings for the node service.
-type NodeOptions struct{}
+type NodeOptions struct {
+	// VolumeAttachLimit specifies the value that shall be reported as "maximum number of attachable volumes"
+	// in CSINode objects. It is similar to https://kubernetes.io/docs/concepts/storage/storage-limits/#custom-limits
+	// which allowed administrators to specify custom volume limits. Note that this solution is just a "stop-gap"
+	// solution until a more sophisticated solution can be implemented (dynamically discovering the maximum number of
```
Comment (gnufied) [inline cmd/options/node_options.go:28]: hmm, I am not sure about documenting this as "stop-gap" where no other solution exists or being worked on. I would drop that wording and document it as such until an alternative presents itself and then we can deprecate this.

File: docs/design.md [hunk 5]
```diff
@@ -235,3 +235,14 @@ Example 2: `AWS_REGION=us-west-1 /bin/aws-ebs-csi-driver controller --extra-volu
...
+This solution is not working with CSI any longer.
+As part of [#347](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/347) we discuss how we can implement a sophisticated computation of the volume attach limit per node (e.g., based on the used machine types or already attached network interfaces).
+However, it turns out that such optimal implementation is not easily achievable. 
+In order to allow existing clusters that are leveraging/relying on this feature to migrate to CSI, the EBS CSI driver is supporting the `--volume-attach-limit` flag.
+It works similar to earlier supported the `KUBE_MAX_PD_VOLS` environment variable.
+If provided, the EBS driver will report the configured value as "maximum number of attachable volumes" in the `CSINode` objects during driver registration. 
+Please consider using this flag as "stop gap" solution that is probably going to be replaced in the future with a more intelligent way of getting to know the attachment limits.
```
Comment (gnufied) [inline docs/design.md:248]: same as above about "stop gap" wording.

File: cmd/options/node_options.go (hunk 4 above)
Comment (rfranzke) [reply cmd/options/node_options.go:28]: Hm, but isn't it exactly like this? It's a stop-gap solution because the alternative would be to not have any possibility to migrate to CSI in case one depends on the Kubernetes feature allowing to overwrite the volume limits. Also, nobody is working on the "real" solution, or? The last update from the ticket is that the easiest implementation would require AWS to expose this information, e.g. via metadata service.

Anyways, what do you suggest to write here instead? Shall I add that the flag is planned to be deprecated in the future when a better solution is implemented?

**Comment by rfranzke** [comment]
Btw, any pointers on how to get the e2e tests fixed? They fail unrelatedly, see https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_aws-ebs-csi-driver/522/pull-aws-ebs-csi-driver-e2e-multi-az/1275303611722633217/build-log.txt.

```
$ curl https://sum.golang.org/lookup/github.com/aws/aws-k8s-tester@v1.3.10-0.20200623051819-1bafc137b062 -I
HTTP/2 410
```

File: cmd/options/node_options.go (hunk 4 above)
Comment (gnufied) [reply cmd/options/node_options.go:28]: I think `KUBE_MAX_PD_VOLS` environment variable should work for CSI drivers too - can you open a github issue in kubernetes/kubernetes repo.

Comment (rfranzke) [reply cmd/options/node_options.go]: @gnufied I've rephrased this part and also the section in `docs/design.md`.

> I think KUBE_MAX_PD_VOLS environment variable should work for CSI drivers too - can you open a github issue in kubernetes/kubernetes repo.

According to @hardikdr it doesn't work, but he is currently reproducing the issue an then will report it to k/k. But yes, independent of this it's still reasonable to have it configurable as you initially suggested in #347.

**Comment by gnufied** [comment]
Looks good to me but we might want to wait for @leakingtapan feedback too. 

/lgtm
/hold

File: cmd/options/node_options.go (hunk 4 above)
Comment (hardikdr) [reply cmd/options/node_options.go]: The env-var `KUBE_MAX_PD_VOLS` doesn't have any effect on max volumes attachments, checked on k8s 1.18 with CSI drivers. But I guess, this should better be handled at the CSI-drivers than scheduler.

File: pkg/driver/node.go [hunk 6]
```diff
@@ -514,6 +516,9 @@ func findNvmeVolume(findName string) (device string, err error) {
 
 // getVolumesLimit returns the limit of volumes that the node supports
 func (d *nodeService) getVolumesLimit() int64 {
+	if d.driverOptions.volumeAttachLimit > 0 {
```
Comment (leakingtapan) [inline pkg/driver/node.go:519]: I think ` d.driverOptions.volumeAttachLimit >= 0` is better, as 0 attach limit is a valid case that ppl might need, and the default -1 will be the option to enable fallback to instance type

**Comment by sjmiller609** [comment]
this is configurable in the helm chart [here](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/271046d1be7507cc21717dbaf96e62d3de235700/charts/aws-ebs-csi-driver/templates/node.yaml#L67)
